### PR TITLE
feat: provide a mechanism to prevent deploying

### LINKED ISSUE #339: Add a way to prevent deploying PRs to stage (e.g. [do not stage] )
┆Issue is synchronized with this [Jira Task](https://mozilla-hub.atlassian.net/browse/CONSVC-1630)
┆epic: Continuous Deployment

### PR DESCRIPTION
**Pull request by Dexterp37**
Fixes #339.

This introduces a mechanism to tag the latest commit in the PR in order to prevent deploying it to stage/prod.
In addition to that, it removes the deprecated release-tagging mechanism (we're now always deploying to stage, so we don't need to tag releases to do that) and documents the current release practices.

I added CI to a sample repo in my account, and here's how it works: [example](https://app.circleci.com/pipelines/github/Dexterp37/test_ci/17/workflows/ffddcb61-259a-4589-ad0f-0b243481b326/jobs/28)

### CODE REVIEW

File: docs/dev/release-process.md [hunk 1]
```diff
@@ -0,0 +1,18 @@
...
+
+Whenever a commit is merged to this repository's `main` branch, the deployment pipeline kicks in, deploying the changeset to the [`stage` environment](../firefox.md).
+After the deployment is complete, accessing the [`__version__` endpoint](https://stage.merino.nonprod.cloudops.mozgcp.net/__version__) will show the commit hash of the deployed version, which will match to the one of the latest commit on the `main` branch.
+
+## Versioning
+The commit hash of the deployed code is considered its version identifier.
+
+## Preventing deployment
+Occasionally developers might want to prevent a commit from triggering the deployment pipeline. While this should be discouraged, there are some legitimate cases for doing so (e.g. docs only changes).
+In order to prevent the deployment of the code from a PR when merging to `main`, the **title of that PR** must contain the `[do not deploy]` text.
```
Comment (ncloudioj) [inline docs/dev/release-process.md:13]: Could you please provide an example commit message here? Also, maybe provide a preferred position to place this tag so that folks can stick to it? For example:

```
# Commit message
doc: Add some docs [do not deploy] 
```

File: docs/dev/release-process.md [hunk 2]
```diff
@@ -0,0 +1,18 @@
+# The Release Process
+
+This project currently follows a [Continuous Delivery](https://en.wikipedia.org/wiki/Continuous_delivery) process, but it's gradually moving toward [Continuous Deployment](https://en.wikipedia.org/wiki/Continuous_deployment).
+
+Whenever a commit is merged to this repository's `main` branch, the deployment pipeline kicks in, deploying the changeset to the [`stage` environment](../firefox.md).
```
Comment (ncloudioj) [inline docs/dev/release-process.md:5]: Nit: I think you can do ```[`stage` environment](../firefox.md#stage)``` to link to the "stage" section.

**Review by ncloudioj — APPROVED** [review]
Looks great, thanks! Just a few nits for your consideration.

File: docs/dev/release-process.md (hunk 1 above)
Comment (Dexterp37) [reply docs/dev/release-process.md:13]: Good idea!

File: docs/dev/release-process.md (hunk 2 above)
Comment (Dexterp37) [reply docs/dev/release-process.md:5]: Oh, correct! Good catch

File: docs/dev/release-process.md [hunk 3]
```diff
@@ -0,0 +1,26 @@
+# The Release Process
+
+This project currently follows a [Continuous Delivery](https://en.wikipedia.org/wiki/Continuous_delivery) process, but it's gradually moving toward [Continuous Deployment](https://en.wikipedia.org/wiki/Continuous_deployment).
+
+Whenever a commit is merged to this repository's `main` branch, the deployment pipeline kicks in, deploying the changeset to the [`stage` environment](../firefox.md#stage).
+After the deployment is complete, accessing the [`__version__` endpoint](https://stage.merino.nonprod.cloudops.mozgcp.net/__version__) will show the commit hash of the deployed version, which will match to the one of the latest commit on the `main` branch.
```
Comment (hackebrot) [inline docs/dev/release-process.md:6]: The following may not always be the case:

> which will match to the one of the latest commit on the `main` branch

Given that the updated CI workflow will skip the deployment if it finds a `[do not deploy]` text in the commit message, the latest commit on the `main` branch may not be deployed to `STAGE`. Also while the deployment is in progress, the `/__version__` may be handled by a node that isn't running the latest version (but that's an edge case).

File: docs/dev/release-process.md [hunk 4]
```diff
@@ -0,0 +1,26 @@
...
+
+## Versioning
+The commit hash of the deployed code is considered its version identifier.
```
Comment (hackebrot) [inline docs/dev/release-process.md:9]: It may be useful to add a comment on how to generate that Git commit hash locally.

File: .circleci/config.yml [hunk 5]
```diff
@@ -294,3 +271,15 @@ commands:
...
+
+  skip-if-do-not-deploy:
+    steps:
+      - run:
+          name: Check if deployment is disallowed
+          # This relies on the [do not deploy] text to be available in the
+          # merge commit when merging the PR to 'main'.
+          command: |
+            if git log -1 "$CIRCLE_SHA1" | grep -q '\[do not deploy\]'; then
+                echo "Skipping this step: deployment was disabled in the PR."
```
Comment (hackebrot) [inline .circleci/config.yml:283]: From what I know `circleci-agent step halt` will **not** terminate the current step. It will also **not** prevent other jobs in the workflow from running.

```suggestion
                echo "Skipping remaining steps in this job: deployment was disabled for this commit."
```

File: docs/dev/release-process.md [hunk 6]
```diff
@@ -0,0 +1,26 @@
...
+
+## Preventing deployment
+Occasionally developers might want to prevent a commit from triggering the deployment pipeline. While this should be discouraged, there are some legitimate cases for doing so (e.g. docs only changes).
+In order to prevent the deployment of the code from a PR when merging to `main`, the **title of that PR** must contain the `[do not deploy]` text. For example:
```
Comment (hackebrot) [inline docs/dev/release-process.md:13]: I want to point out that the GitHub web UI allows to write a custom commit message for merge commits including the optional extended description which by default contains the pull request title. It may be worth calling out that this must not be changed or that `[do not deploy]` needs to be included in the commit message if desired.

**Review by hackebrot — CHANGES_REQUESTED** [review]
Hi @Dexterp37! 👋🏻 

I added my comments inline.

File: docs/dev/release-process.md [hunk 7]
```diff
@@ -0,0 +1,26 @@
+# The Release Process
+
+This project currently follows a [Continuous Delivery](https://en.wikipedia.org/wiki/Continuous_delivery) process, but it's gradually moving toward [Continuous Deployment](https://en.wikipedia.org/wiki/Continuous_deployment).
+
+Whenever a commit is merged to this repository's `main` branch, the deployment pipeline kicks in, deploying the changeset to the [`stage` environment](../firefox.md#stage).
```
[the remark this answers is not in the document]
Comment (Dexterp37) [reply docs/dev/release-process.md:5]: I believe commit is the right word here, because the deployment pipeline kicks in even if you directly commit to the main branch without a the commit coming from another branch. Or did you mean something else, @hackebrot ?

File: docs/dev/release-process.md (hunk 3 above)
Comment (Dexterp37) [reply docs/dev/release-process.md:6]: The sentence starts with `After the deployment is complete`, so I'm assuming that deployment happened in the rest of the sentence. Happy to accept suggestions on making this clearer if it is uncear, though.

Good point on the version part, I can add that it will "eventually" match given that the endpoint could be served by an older version still living.

File: docs/dev/release-process.md (hunk 4 above)
Comment (Dexterp37) [inline docs/dev/release-process.md:9]:
```suggestion
The commit hash of the deployed code is considered its version identifier. The commit hash can be retrieved locally via `git rev-parse HEAD`.
```

File: docs/dev/release-process.md (hunk 6 above)
Comment (Dexterp37) [inline docs/dev/release-process.md:13]:
```suggestion
In order to prevent the deployment of the code from a PR when merging to `main`, the **title of that PR** must contain the `[do not deploy]` text. Note that, when generating the merge commit for a branch within the GitHub UI, the extened description must not be changed or care must be taken to ensure that `[do not deploy]`  is still present.
For example:
```

File: docs/dev/release-process.md (hunk 7 above)
[the remark this answers is not in the document]
Comment (hackebrot) [reply docs/dev/release-process.md:8]: That makes sense. I would rephrase to

> Whenever a commit is pushed to this repository's `main` branch.

in that case, because you cannot "merge" a single Git commit.

File: docs/dev/release-process.md [hunk 8]
```diff
@@ -0,0 +1,32 @@
+# The Release Process
+
+This project currently follows a [Continuous Delivery][continuous_delivery] process, but it's gradually moving toward [Continuous Deployment][continuous_deployment].
+
+[continuous_delivery]: https://en.wikipedia.org/wiki/Continuous_delivery
+[continuous_deployment]: https://en.wikipedia.org/wiki/Continuous_deployment
+
+Whenever a commit is merged to this repository's `main` branch, the deployment pipeline kicks in, deploying the changeset to the [`stage` environment](../firefox.md#stage).
```
Comment (Dexterp37) [inline docs/dev/release-process.md:8]:
```suggestion
Whenever a commit is pushed to this repository's `main` branch, the deployment pipeline kicks in, deploying the changeset to the [`stage` environment](../firefox.md#stage).
```

**Review by hackebrot — APPROVED** [review]
LGTM! 🚀

**Comment by hackebrot** [comment]
Looks like not all commits on this branch are signed and merging is blocked as a result.
